cfgmgr/portmgr.cpp

```diff
@@ -23,27 +23,55 @@ PortMgr::PortMgr(DBConnector *cfgDb, DBConnector *appDb, DBConnector *stateDb, c
 bool PortMgr::setPortMtu(const string &alias, const string &mtu)
 {
     stringstream cmd;
-    string res;
+    string res, cmd_str;
 
     // ip link set dev <port_name> mtu <mtu>
     cmd << IP_CMD << " link set dev " << shellquote(alias) << " mtu " << shellquote(mtu);
-    EXEC_WITH_ERROR_THROW(cmd.str(), res);
-
-    // Set the port MTU in application database to update both
-    // the port MTU and possibly the port based router interface MTU
-    return writeConfigToAppDb(alias, "mtu", mtu);
+    cmd_str = cmd.str();
+    int ret = swss::exec(cmd_str, res);
+    if (!ret)
+    {
+        // Set the port MTU in application database to update both
+        // the port MTU and possibly the port based router interface MTU
+        return writeConfigToAppDb(alias, "mtu", mtu);
+    }
+    else if (!isPortStateOk(alias))
+    {
+        // Can happen when a DEL notification is sent by portmgrd followed by a pending mtu SET notification
+        SWSS_LOG_WARN("Setting mtu to alias:%s netdev failed with cmd:%s, rc:%d, error:%s", alias.c_str(), cmd_str.c_str(), ret, res.c_str());
+        return false;
+    }
+    else
+    {
+        throw runtime_error(cmd_str + " : " + res);
+    }
+    return true;
 }
 
 bool PortMgr::setPortAdminStatus(const string &alias, const bool up)
 {
     stringstream cmd;
-    string res;
+    string res, cmd_str;
 
     // ip link set dev <port_name> [up|down]
     cmd << IP_CMD << " link set dev " << shellquote(alias) << (up ? " up" : " down");
-    EXEC_WITH_ERROR_THROW(cmd.str(), res);
-
-    return writeConfigToAppDb(alias, "admin_status", (up ? "up" : "down"));
+    cmd_str = cmd.str();
+    int ret = swss::exec(cmd_str, res);
+    if (!ret)
+    {
+        return writeConfigToAppDb(alias, "admin_status", (up ? "up" : "down"));
+    }
+    else if (!isPortStateOk(alias))
+    {
+        // Can happen when a DEL notification is sent by portmgrd followed by a pending admin_status SET notification
```

Comment:
Not sure I understand the comment, could you please elaborate?

Reply:
portmgrd defers adds the admin_status/mtu notifs back to the consumer queue when the netdev is not created. So, let's assume, there is a DEL notif which arrived after the first SET, We'll process DEL first and then followed by the deferred admin_status/mtu notif. 

If this happens there might be a possibility where portmgrd tries to set mtu/admin_status on a deleted netdev iface and in process throws a runtime error.

Reply:
I am not sure that if it is possible. I suppose DEL will override everything.  Could you please check this https://github.com/sonic-net/sonic-swss/blob/ac8a83f0cd36c11c1a480a682e4eabbd99bf962c/orchagent/orch.cpp#L100 ?

Reply:
Oh, i'm sorry. I was interpreting it wrong.  

A DEL came first and before the STATE_DB got updated, a SET came with all the fields.
```
2022-09-06.01:30:50.803955|PORT_TABLE:Ethernet64|DEL
2022-09-06.01:30:50.841680|PORT_TABLE:Ethernet64|SET|alias:fortyGigE0/64|index:16|lanes:69,70,71,72|speed:100000
2022-09-06.01:30:51.854363|PORT_TABLE:Ethernet64|SET|mtu:9100
2022-09-06.01:30:51.860562|PORT_TABLE:Ethernet64|SET|admin_status:down
```

But the scenario certainly happened: This was the log in one of the tests:
```
2022-09-01T09:22:37.2380593Z ______________________ TestPortDPB.test_port_breakout_one ______________________
2022-09-01T09:22:37.2381024Z 
2022-09-01T09:22:37.2381526Z self = <test_port_dpb.TestPortDPB object at 0x7f90ce7d0850>
2022-09-01T09:22:37.2382156Z dvs = <conftest.DockerVirtualSwitch object at 0x7f90ce82fe80>
2022-09-01T09:22:37.2382620Z 
2022-09-01T09:22:37.2383089Z     def test_port_breakout_one(self, dvs):
2022-09-01T09:22:37.2383573Z         dpb = DPB()
2022-09-01T09:22:37.2384079Z >       dpb.breakout(dvs, "Ethernet0", maxBreakOut)
2022-09-01T09:22:37.2384424Z 
2022-09-01T09:22:37.2384862Z test_port_dpb.py:45: 
2022-09-01T09:22:37.2385414Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2022-09-01T09:22:37.2385981Z port_dpb.py:304: in breakout
2022-09-01T09:22:37.2386534Z     self.create_child_ports(dvs, p, num_child_ports)
2022-09-01T09:22:37.2387143Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2022-09-01T09:22:37.2387503Z 
2022-09-01T09:22:37.2387967Z self = <port_dpb.DPB object at 0x7f90cece7af0>
2022-09-01T09:22:37.2388571Z dvs = <conftest.DockerVirtualSwitch object at 0x7f90ce82fe80>
2022-09-01T09:22:37.2389216Z p = <port_dpb.Port object at 0x7f90cece7730>, num_child_ports = 4
2022-09-01T09:22:37.2389605Z 
2022-09-01T09:22:37.2390103Z     def create_child_ports(self, dvs, p, num_child_ports):
2022-09-01T09:22:37.2390697Z         # Create child ports and write to config DB
2022-09-01T09:22:37.2391263Z         child_ports = p.port_split(num_child_ports)
2022-09-01T09:22:37.2391792Z         child_port_names = []
2022-09-01T09:22:37.2392263Z         for cp in child_ports:
2022-09-01T09:22:37.2392754Z             cp.write_to_config_db()
2022-09-01T09:22:37.2393279Z             child_port_names.append(cp.get_name())
2022-09-01T09:22:37.2393897Z         #print "Added child ports:%s to config DB"%child_port_names
2022-09-01T09:22:37.2394413Z     
2022-09-01T09:22:37.2394843Z         for cp in child_ports:
2022-09-01T09:22:37.2395361Z             assert(cp.exists_in_config_db() == True)
2022-09-01T09:22:37.2395889Z             cp.verify_config_db()
2022-09-01T09:22:37.2396501Z         #print "Config DB verification passed"
2022-09-01T09:22:37.2397019Z         for cp in child_ports:
2022-09-01T09:22:37.2397536Z >           assert(cp.exists_in_app_db() == True)
2022-09-01T09:22:37.2398045Z E           AssertionError
2022-09-01T09:22:37.2398333Z 
2022-09-01T09:22:37.2398778Z port_dpb.py:279: AssertionError
```

This happened for first child port, portmgrd crashed and thus nothing was written to APP_DB for other child ports. i've checked in my local vs and saw the portmgrd runtime error log as well